packages/server/src/fhir/repo.ts

```diff
@@ -1906,32 +1907,58 @@ export class Repository extends BaseRepository implements FhirRepository<PoolCli
    */
   getDatabaseClient(): Pool | PoolClient {
     this.assertNotClosed();
-    return getDatabasePool();
+    // If in a transaction, then use the transaction client.
+    // Otherwise, use the pool client.
+    return this.conn ?? getDatabasePool();
+  }
+
+  /**
+   * Returns a proper database connection.
+   * Unlike getDatabaseClient(), this method always returns a PoolClient.
+   * @returns Database connection.
+   */
+  private async getConnection(): Promise<PoolClient> {
+    this.assertNotClosed();
+    if (!this.conn) {
+      this.conn = await getDatabasePool().connect();
+    }
+    return this.conn;
+  }
+
+  /**
+   * Releases the database connection.
+   * Include an error to remove the connection from the pool.
+   * See: https://github.com/brianc/node-postgres/blob/master/packages/pg-pool/index.js#L333
+   * @param err - Optional error to remove the connection from the pool.
+   */
+  private releaseConnection(err?: boolean | Error): void {
+    if (this.conn) {
+      this.conn.release(err);
+      this.conn = undefined;
+    }
   }
 
   async withTransaction<TResult>(callback: (client: PoolClient) => Promise<TResult>): Promise<TResult> {
-    const conn = await getDatabasePool().connect();
     try {
-      await conn.query('BEGIN');
-      const result = await callback(conn);
-      await conn.query('COMMIT');
-      conn.release();
+      const client = await this.beginTransaction();
+      const result = await callback(client);
+      await this.commitTransaction();
       return result;
     } catch (err: any) {
-      globalLogger.error('Transaction error', err);
       const operationOutcomeError = new OperationOutcomeError(normalizeOperationOutcome(err), err);
-      try {
-        await conn.query('ROLLBACK');
-      } catch (err2: any) {
-        globalLogger.error('Rollback error', err2);
-      }
-      conn.release(err);
+      await this.rollbackTransaction(operationOutcomeError);
       throw operationOutcomeError;
+    } finally {
+      this.endTransaction();
     }
   }
 
   close(): void {
     this.assertNotClosed();
+    this.releaseConnection();
+    if (this.transactionDepth > 0) {
+      getRequestContext().logger.error('Closing Repository with active transaction');
+    }
```

Comment:
Minor suggestion, probably paranoia, for your consideration

```suggestion
    if (this.transactionDepth > 0) {
      // Bad state, remove connection from pool
      getRequestContext().logger.error('Closing Repository with active transaction');
      this.releaseConnection(new Error('Closing Repository with active transaction'));
    } else {
      // Good state, return healthy connection to pool
      this.releaseConnection();
    }
```